subsys/fs/ext2/ext2_diskops.c

```diff
@@ -0,0 +1,906 @@
+			 * single blocks, hence we will just remove them.
+			 * We can just set start_blk here and remove blocks in loop at the end of
+			 * this function.
+			 */
+			start_blk = offsets[0];
+
+		} else {
+			/* We don't remove all blocks referenced by current block. We have to use
+			 * offsets to decide which part of next block we want to remove.
+			 */
+			__ASSERT(list[offsets[0]] != 0,
+					"Block that references some other must be nonzero");
+
+			/* We will start removing whole blocks from next block on this level */
+			start_blk = offsets[0] + 1;
+
+			/* Remove desired part of lower level block. */
+			rem = delete_blocks(fs, list[offsets[0]], lvl - 1, &offsets[1]);
+			if (rem < 0) {
+				removed = rem;
+				goto out;
+			}
+			removed += rem;
+		}
+
+		/* Iterate over blocks that will be entirely deleted */
+		for (uint32_t i = start_blk; i < fs->block_size / sizeof(uint32_t); ++i) {
+			if (list[i] == 0) {
+				continue;
+			}
+			rem = delete_blocks(fs, list[i], lvl - 1, zero_offsets);
+			if (rem < 0) {
+				removed = rem;
+				goto out;
+			}
+			removed += rem;
+			list[i] = 0;
+			block_dirty = true;
+		}
+	}
+
+	if (remove_current) {
+		LOG_DBG("free block %d (lvl %d)", block_num, lvl);
+
+		/* If we remove current block, we don't have to write it's updated content. */
+		if (list_block) {
+			block_dirty = false;
+		}
+
+		ret = ext2_free_block(fs, block_num);
+		if (ret < 0) {
+			removed = ret;
+		}
+	}
+out:
+	if (removed >= 0 && list_block && block_dirty) {
+		ret = ext2_write_block(fs, list_block);
+		if (ret < 0) {
+			removed = ret;
+		}
+	}
+	ext2_drop_block(list_block);
+
+	/* On error removed will contain negative error code */
+	return removed;
+}
+
+static int get_level_offsets(struct ext2_data *fs, uint32_t block, uint32_t offsets[4])
+{
+	const uint32_t B = fs->block_size / sizeof(uint32_t);
+	const uint32_t lvl0_blks = 12;
+	const uint32_t lvl1_blks = B;
+	const uint32_t lvl2_blks = B * B;
+	const uint32_t lvl3_blks = B * B * B;
+
+	if (block < 0) {
+		return -EINVAL;
+	}
+
+	/* Level 0 */
+	if (block < lvl0_blks) {
+		offsets[0] = block;
+		return 0;
+	}
+
+	/* Level 1 */
+	block -= lvl0_blks;
+	if (block < lvl1_blks) {
+		offsets[0] = 12;
+		 offsets[1] = block;
+		return 1;
+	}
+
+	/* Level 2 */
+	block -= lvl1_blks;
+	if (block < lvl2_blks) {
+		offsets[0] = 13;
+		offsets[1] = block / B;
+		offsets[2] = block % B;
+		return 2;
+	}
+
+	/* Level 3 */
+	if (block < lvl3_blks) {
+		block -= lvl2_blks;
+		offsets[0] = 13;
+		offsets[1] = block / (B * B);
+		offsets[2] = (block % (B * B)) / B;
+		offsets[3] = (block % (B * B)) % B;
+		return 3;
+	}
+	/* Block number is too large */
+	return -EINVAL;
+}
+
+static int block0_level(uint32_t block)
+{
+	if (block < EXT2_INODE_BLOCK_DIRECT) {
+		return 0;
+	} else if (block  == EXT2_INODE_BLOCK_1LVL) {
+		return 1;
+	} else if (block  == EXT2_INODE_BLOCK_2LVL) {
+		return 2;
+	} else if (block  == EXT2_INODE_BLOCK_3LVL) {
+		return 3;
+	}
+	return -1;
```

Comment:
If failure is not possible, so `block` is always correct and in set of allowed values), this can be shortened to
```
if (block >=  EXT2_INODE_BLOCK_1LVL)  {
  return (block ^ EXT2_INODE_BLOCK_1LVL) + 1;
}
return 0;
```

Reply:
You're right. Failure here is not possible.